### PR TITLE
First part of rewrite to use new Bitcode Types

### PR DESCRIPTION
**Pull request by pointhi**
* Add adapter between old and new type
* rewrite LLVMLiteralFactory to use new type
* add getBytes methode as done for the old type, to simplify code

Old (outdated) PR:  #484

### CODE REVIEW

File: projects/com.oracle.truffle.llvm.parser.factories/src/com/oracle/truffle/llvm/parser/factories/LLVMAggregateFactory.java [hunk 1]
```diff
@@ -113,29 +114,29 @@ public static LLVMExpressionNode createInsertValue(LLVMAddressNode resultAggrega
...
-        LLVMExpressionNode alloc = runtime.allocateFunctionLifetime(structType, structSize, structAlignment);
+        int structSize = ((LLVMTypeHelperImpl) runtime.getTypeHelper()).getByteSize(structType);
+        int structAlignment = ((LLVMTypeHelperImpl) runtime.getTypeHelper()).getAlignmentByte(structType);
+        LLVMExpressionNode alloc = runtime.allocateFunctionLifetime(LLVMToBitcodeAdapter.unresolveType(structType), structSize, structAlignment);
         for (int i = 0; i < types.length; i++) {
-            ResolvedType resolvedType = types[i];
+            Type resolvedType = types[i];
             if (!packed) {
-                currentOffset += runtime.getTypeHelper().computePaddingByte(currentOffset, resolvedType);
+                currentOffset += ((LLVMTypeHelperImpl) runtime.getTypeHelper()).computePaddingByte(currentOffset, resolvedType);
```
Comment (mrigger) [inline projects/com.oracle.truffle.llvm.parser.factories/src/com/oracle/truffle/llvm/parser/factories/LLVMAggregateFactory.java]: I saw that you introduced a `LLVMTypeHelper` interface but still cast it to `LLVMTypeHelperImpl` in some instances. Can we not just remove the interface and use the implementation class directly?

Comment (jkreindl) [inline projects/com.oracle.truffle.llvm.parser.factories/src/com/oracle/truffle/llvm/parser/factories/LLVMAggregateFactory.java]: I agree that we do not need the Interface. As dir the Rest of the changes, I will Take a Look at them this afternoon, @mrigger could you please Hold Off on merging until then?

Comment (mrigger) [inline projects/com.oracle.truffle.llvm.parser.factories/src/com/oracle/truffle/llvm/parser/factories/LLVMAggregateFactory.java]: Ok.

File: projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        }
+
+        throw new AssertionError("Unknown type: " + type + " - " + type.getClass().getTypeName());
+    }
+
+    private static Map<ResolvedNamedType, Type> namedTypes = new HashMap<>();
+
+    public static Type resolveType(ResolvedNamedType type) {
+        if (!namedTypes.containsKey(type)) {
+            namedTypes.put(type, MetaType.UNKNOWN); // TODO: resolve cycles
```
Comment (jkreindl) [inline projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java]: Do we really need both puts? The second would overwrite the first anyways.

File: projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        for (ResolvedType arg : type.getParameters()) {
+            assert !hasVararg; // should be the last element of the parameterlist
+            if (arg.isVararg()) {
+                hasVararg = true;
+            } else {
+                args.add(resolveType(arg));
+            }
+        }
+        FunctionType fType = new FunctionType(returnType, args.toArray(new Type[args.size()]), hasVararg);
+        return new FunctionDeclaration(fType);
```
Comment (jkreindl) [inline projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java]: Should we distinguish between FunctionDeclaration and FunctionDefinition, or is this enough for the NodeFactoryFacade?

File: projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    private static final int HALF_SIZE = 16;
```
Comment (jkreindl) [inline projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java]: These constants should really be public ones in the FloatingPointType Class.

File: projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java (hunk 2 above)
Comment (jkreindl) [inline projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java]: Nevermind, I see that the resolveType call in the second put might access the value stored in the first.

File: projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java (hunk 4 above)
Comment (pointhi) [inline projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java]: done for FloatingPoint and Integer types (https://github.com/graalvm/sulong/pull/489/commits/98db09f9a6e9325475c0b5002d182313c1b35ca5)

File: projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java (hunk 3 above)
Comment (pointhi) [inline projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java]: I didn't found a representation of FunctionDefinition in llvm.ireditor, so I haven't implemented it

File: projects/com.oracle.truffle.llvm.parser.factories/src/com/oracle/truffle/llvm/parser/factories/LLVMAggregateFactory.java (hunk 1 above)
Comment (pointhi) [inline projects/com.oracle.truffle.llvm.parser.factories/src/com/oracle/truffle/llvm/parser/factories/LLVMAggregateFactory.java]: Done.

This was a temporary solution to solve dependency problems across packages (https://github.com/graalvm/sulong/pull/489/commits/29a45d753fe4b6fa4208212ec55eafb5ff717e6d)

**Comment by pointhi** [comment]
Most parts of this PR were merged:  #499 

The only commit which could be still interesting is: https://github.com/graalvm/sulong/pull/489/commits/98db09f9a6e9325475c0b5002d182313c1b35ca5

**Pull request closed without merging**